File: whatsapp/bakery/src/main/java/com/tranche/bakery/whatsapp/WhatsAppClient.java
package com.tranche.bakery.whatsapp;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

@Service
@Slf4j
public class WhatsAppClient {

    private final RestClient restClient;
    private final String phoneNumberId;

    public WhatsAppClient(
            @Value("${whatsapp.api.url}") String apiUrl,
            @Value("${whatsapp.api.token}") String token,
            @Value("${whatsapp.api.phone-number-id}") String phoneNumberId) {
        this.phoneNumberId = phoneNumberId;
        this.restClient = RestClient.builder()
                .baseUrl(apiUrl)
                .defaultHeader("Authorization", "Bearer " + token)
                .defaultHeader("Content-Type", MediaType.APPLICATION_JSON_VALUE)
                .build();
    }

    public void sendText(String to, String body) {
        send(WhatsAppMessage.text(to, body));
    }

    public void sendButtons(String to, String bodyText, java.util.List<WhatsAppMessage.Button> buttons) {
        send(WhatsAppMessage.buttonMessage(to, bodyText, buttons));
    }

    public void sendList(String to, String bodyText, String buttonLabel, java.util.List<WhatsAppMessage.Section> sections) {
        send(WhatsAppMessage.listMessage(to, bodyText, buttonLabel, sections));
    }

    private void send(Object message) {
        try {
            String response = restClient.post()
                    .uri("/{phoneNumberId}/messages", phoneNumberId)
                    .body(message)
                    .retrieve()
                    .body(String.class);
            log.debug("WhatsApp API response: {}", response);
        } catch (Exception e) {
            log.error("Failed to send WhatsApp message: {}", e.getMessage());
        }
    }
}

File: whatsapp/temp/src/services/whatsappClient.js
export class WhatsAppClient {
  constructor(config) {
    this.config = config;
  }

  async sendText(to, body) {
    if (!this.config.accessToken || !this.config.phoneNumberId) {
      console.log(`[dev whatsapp] -> ${to}: ${body}`);
      return { dev: true };
    }

    const url = `https://graph.facebook.com/${this.config.apiVersion}/${this.config.phoneNumberId}/messages`;
    const response = await fetch(url, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.config.accessToken}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        messaging_product: "whatsapp",
        recipient_type: "individual",
        to,
        type: "text",
        text: { preview_url: false, body }
      })
    });

    if (!response.ok) {
      const errorBody = await response.text();
      throw new Error(`WhatsApp send failed: ${response.status} ${errorBody}`);
    }

    return response.json();
  }
}
